File: cellular_automata.py
# This script will create Stephen Wolfram's elementary cellular automata
# /home/alan/.local/bin not on path!


import sys
import random
import numpy as np
from PIL import Image


# Returns a numpy array of the given size
# with the middlemost bit turned on
def createBoard(size):
    if size % 2 == 0:
        size += 1
    board = np.zeros(size)
    board[int(len(board)/2)] = 1
    return board



# Prints a slice of the board
def printBoard(board):
    for i in range(len(board)):
        if np.all(board[i], 0):
            print("█", end=' ')
        else:
            print("-", end=' ')
    print()



# Outputs a png image of the full board
def createImage(master):
    img = Image.new('RGB', (len(master), len(master[0])), color = 'white')
    pixels = img.load()
    for i in range(len(master)):
        for j in range(len(master[0])):
            if master[i][j] == 0:
                pixels[i, j] = (0, 0, 0)
            else:
                pixels[i, j] = (255, 255, 255)
    # Needs flipped across the veritcle axis and rotated 90 degrees clockwise
    img = img.transpose(Image.FLIP_TOP_BOTTOM)
    img = img.transpose(Image.ROTATE_270)
    img.save('my.png')
    img.show()



# Defines the type of neighborhood
# and returns a int 0-7
def neighborhood(board, position):
    slice = board[position -1 : position + 2]
    if np.array_equal(slice, np.array([0, 0, 0])):
        return 0
    if np.array_equal(slice, np.array([0, 0, 1])):
        return 1
    if np.array_equal(slice, np.array([0, 1, 0])):
        return 2
    if np.array_equal(slice, np.array([0, 1, 1])):
        return 3
    if np.array_equal(slice, np.array([1, 0, 0])):
        return 4
    if np.array_equal(slice, np.array([1, 0, 1])):
        return 5
    if np.array_equal(slice, np.array([1, 1, 0])):
        return 6
    if np.array_equal(slice, np.array([1, 1, 1])):
        return 7



# Turns a decimal number into an binary representation,
# stores that in an array, and reverses the order
def applyRule(ruleNumber):
    rule = np.binary_repr(ruleNumber)
    rule = rule.zfill(8)
    rule = np.array(list(rule))
    rule = [int(i) for i in rule]
    rule.reverse()
    return rule



# Returns a the next state of the board 
def step(board, rule):
    new_board = createBoard(len(board))
    rule = applyRule(rule)
    for i in range(1, len(board) - 1):
        n = neighborhood(board, i)
        new_board[i] = rule[n]
    return new_board


"""
Primary loop. Appends each slice(type numpy array)
and appends it to the master list. After the given
number of iterations, a full image will be created.
The bottom-most for loop can be used instead of the
createImage function to simply print the full board
to the terminal.
"""
def loop(size, rule, iterations):
    board = createBoard(size)
    master = []
    
    master.append(board)
    printBoard(master)
    for i in range(iterations):
        board = step(board, rule)
        master.append(board)
    #for i in range(len(master)):
    #    printBoard(master[i])
    createImage(master)


"""
Callable from command line
Args  | type | Description
-----------------------------------------------------
@size | int  | How many pixels wide the board will be
@rule | int  | Which rule number to apply, 0-255 valid
"""
def main():
    arg1 = sys.argv[1]
    arg2 = sys.argv[2]

    # Input error handlng
    if rule < 0 or rule > 255:
        print("Error - Rule must be between 0 and 255")
        exit()
    if rule % 1 != 0:
        print("Error - Rule must be an integer")
        exit()

    #size = iterations(2) - 1
    #(size + 1) / 2 = iterations
    size = int(sys.argv[1])
    rule = int(sys.argv[2])
    iterations = round((size + 1) /2)

    loop(size, rule, iterations)


#main()

test = float(sys.argv[1])
print(type(test))

File: cellular_automata.go
// Create elementary cellular automata given a rule number
package main

import (
	"fmt"
)

// Currently, this function prints the board to 
// the console. Modify it to use a graphics library
func printBoard(board []int) {
	for i := 0; i < len(board); i++ {
		fmt.Printf("%d", board[i])
	}
	fmt.Println()
}


func decimalToBinary(n int) int {
	// convert a decimal number to binary
	// return the binary number

	dec := n 
}




// return the left, center, and right cells
// for a given cell in the board
//	// if the cell is on the edge of the board, 
//	// the edge cells wrap around to the other side of the board
func neighborhood(board []int, position int) (int, int, int) {

	switch board[position - 1 : position + 2] {
		case [0, 0, 0]:
			return 7
		case [0, 0, 1]:
			return 6
		case [0, 1, 0]:
			return 5
		case [0, 1, 1]:
			return 4
		case [1, 0, 0]:
			return 3
		case [1, 0, 1]:
			return 2
		case [1, 1, 0]:
			return 1
		case [1, 1, 1]:			
	}
}

	// create a 1D slice of 0s and 1s
	// for each cell in the slice, apply the rule to the cell and its neighbors
	// print the slice

	// create a slice of 0s and 1s
	// for each cell in the slice, apply the rule to the cell and its neighbors
	// print the slice

func generation(board []int, rule int) []int {
	// create a new slice of the same length as the board
	// for each cell in the slice, apply the rule to the cell and its neighbors
	// return the new slice

	newBoard := make([]int, len(board))
	for i := 0; i < len(board); i++ {
		newBoard[i] = neighborhood(board, i)
	}
}


func createCa(int rule, int length){
	board := make([]int, length)
	for i := 0; i < length; i++ {
		board[i] = 0
	}

	// set the middle cell to 1
	board[length/2] = 1

	printBoard(board)

	for i := 0; i < length; i++ {
		// apply the rule to the cell and its neighbors
		// print the slice
	}



}

func main() {

}
